File: app/src/main/java/us/shandian/giga/postprocessing/M4aNoDash.java
package us.shandian.giga.postprocessing;

import org.schabi.newpipe.streams.Mp4DashReader;
import org.schabi.newpipe.streams.Mp4FromDashWriter;
import org.schabi.newpipe.streams.io.SharpStream;

import java.io.IOException;

class M4aNoDash extends Postprocessing {

    M4aNoDash() {
        super(false, true, ALGORITHM_M4A_NO_DASH);
    }

    @Override
    boolean test(SharpStream... sources) throws IOException {
        // check if the mp4 file is DASH (youtube)

        Mp4DashReader reader = new Mp4DashReader(sources[0]);
        reader.parse();

        switch (reader.getBrands()[0]) {
            case 0x64617368:// DASH
            case 0x69736F35:// ISO5
                return true;
            default:
                return false;
        }
    }

    @Override
    int process(SharpStream out, SharpStream... sources) throws IOException {
        Mp4FromDashWriter muxer = new Mp4FromDashWriter(sources[0]);
        muxer.setMainBrand(0x4D344120);// binary string "M4A "
        muxer.parseSources();
        muxer.selectTracks(0);
        muxer.build(out);

        return OK_RESULT;
    }
}

File: app/src/main/java/us/shandian/giga/postprocessing/M4aNoDash.kt
package us.shandian.giga.postprocessing

import okio.IOException
import org.schabi.newpipe.streams.Mp4DashReader
import org.schabi.newpipe.streams.Mp4FromDashWriter
import org.schabi.newpipe.streams.io.SharpStream

internal class M4aNoDash : Postprocessing(false, true, ALGORITHM_M4A_NO_DASH) {
    override suspend fun test(vararg sources: SharpStream): Boolean {
        val reader = Mp4DashReader(sources[0])
        reader.parse()

        val brands = reader.getBrands()
        return when (if (brands.isNotEmpty()) brands[0] else 0) {
            0x64617368, // DASH
            0x69736F35 -> true

            // ISO5
            else -> false
        }
    }

    override suspend fun process(out: SharpStream?, vararg sources: SharpStream): Int {
        val muxer = Mp4FromDashWriter(sources[0])
        muxer.setMainBrand(0x4D344120) // binary string "M4A "
        muxer.parseSources()
        muxer.selectTracks(0)
        muxer.build(out as SharpStream)

        return OK_RESULT.toInt()
    }
}
